PennMobile/src/main/java/com/pennapps/labs/pennmobile/DiningHallWidget.kt

```diff
@@ -0,0 +1,93 @@
+package com.pennapps.labs.pennmobile
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.widget.RemoteViews
+import com.google.gson.GsonBuilder
+import com.google.gson.reflect.TypeToken
+import com.pennapps.labs.pennmobile.adapters.DiningHallWidgetAdapter
+import com.pennapps.labs.pennmobile.api.DiningRequest
+import com.pennapps.labs.pennmobile.api.Serializer
+import com.pennapps.labs.pennmobile.classes.DiningHall
+import com.pennapps.labs.pennmobile.classes.Venue
+import com.squareup.okhttp.OkHttpClient
+import retrofit.RestAdapter
+import retrofit.client.OkClient
+import retrofit.converter.GsonConverter
+import java.util.concurrent.TimeUnit
+
+/**
+ * Implementation of App Widget functionality.
+ */
+class DiningHallWidget : AppWidgetProvider() {
+    override fun onUpdate(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetIds: IntArray
+    ) {
+        for (appWidgetId in appWidgetIds) {
+            val mainActivityIntent = Intent(context, MainActivity::class.java)
+            mainActivityIntent.apply {
+                flags += Intent.FLAG_ACTIVITY_NEW_TASK
+                flags += Intent.FLAG_ACTIVITY_CLEAR_TOP
+            }
+            mainActivityIntent.putExtra("Widget_Tab_Switch", 2)
+            val serviceIntent = Intent(context, DiningHallWidgetAdapter::class.java)
+            val pendingIntent : PendingIntent = PendingIntent.getActivity(context, appWidgetId, mainActivityIntent, PendingIntent.FLAG_IMMUTABLE)
+            serviceIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+            serviceIntent.setData(Uri.parse(serviceIntent.toUri(Intent.URI_INTENT_SCHEME)))
+            val views = RemoteViews(context.packageName, R.layout.dining_hall_widget)
+            views.setRemoteAdapter(R.id.stackview, serviceIntent)
+            views.setEmptyView(R.id.stackview, R.id.emptyview)
+            views.setPendingIntentTemplate(R.id.stackview, pendingIntent)
+            views.setOnClickPendingIntent(R.id.emptyview, pendingIntent)
+            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.stackview)
+            appWidgetManager.updateAppWidget(appWidgetId, views)
+        }
+    }
+
+    override fun onEnabled(context: Context) {
```

Comment:
is it normal for these functions to be empty?

Reply:
It's normal for these functions to be empty. The main purpose of these functions are mainly used for AlarmManager testing (but from most of the example projects they are just not used at all).